python/arcticdb/adapters/prefixing_library_adapter_decorator.py

```diff
@@ -0,0 +1,82 @@
+"""
+Copyright 2023 Man Group Operations Limited
+
+Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+
+As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+"""
+import inspect
+import re
+from typing import Optional, Tuple, Type, Callable, TYPE_CHECKING, Iterable, List
+
+from arcticdb import LibraryOptions, Arctic
+from arcticdb.adapters.arctic_library_adapter import ArcticLibraryAdapter
+from arcticdb.encoding_version import EncodingVersion
+
+
+class PrefixingLibraryAdapterDecorator(ArcticLibraryAdapter if TYPE_CHECKING else object):
```

Comment:
I find this quite confusing, because:

- It is only used for Mongo, but that isn't obvious
- The `__getattr__` wrapper is complex
- The test fixture modifying `Arctic._LIBRARY_ADAPTERS` is surprising
- I don't understand the `PrefixingLibraryAdapterDecoratorChecked` concept

Why not implement the prefixing feature for Mongo directly in the non-test tree by modifying `mongo_library_adapter.py`? We can keep the prefix feature undocumented and warn if it is used so that we don't commit to it as an API feature.

Reply:
>It is only used for Mongo, but that isn't obvious
It _can_ be used for any storage type. I have plans to use it for LMDB to address the performance problem of repeatedly creating new storage on Windows.

>The __getattr__ wrapper is complex
This is for safety.

For example, say someone added (or changed) a method in `ArcticLibraryAdapter` that takes a library name and s/he didn't know this prefixing thing, then the real library name will "leak" in that function, which can be a very confusing bug.

This has actually happened twice since I started this work, which is why I wrote this.

>The test fixture modifying Arctic._LIBRARY_ADAPTERS is surprising
How so? I would expect in the future, with some modularization, all LibraryAdapters would be dynamically registered this way.

>I don't understand the PrefixingLibraryAdapterDecoratorChecked concept
It's a marker attribute to avoid having to inspect the called functions each and every time. Computing the `Signature` of a function is quite slow.

>Why not implement the prefixing feature for Mongo directly in the non-test tree by modifying mongo_library_adapter.py?
It's
* conceptually cleaner (using a standard, self-contained decorator pattern)
* more reusable (see my reply to your first point)
* safer (similar to the check in `__getattr__`, if someone added a new method in `ArcticLibraryAdapter` with a default implementation and didn't know the mongo adapter has a prefixing logic in it, then the unprefixed name will leak through that default implementation)